File: 2. Medium/622. Design Circular Queue/solution.js
/**
 * @param {number} k
 */
var MyCircularQueue = function (k) {
    this.head = 0;
    this.tail = -1;
    this.size = 0;
    this.data = new Array(k);
};

/** 
 * @param {number} value
 * @return {boolean}
 */
MyCircularQueue.prototype.enQueue = function (value) {
    if (this.isFull()) {
        return false;
    }

    this.tail = ++this.tail % this.data.length;
    this.data[this.tail] = value;
    this.size++;

    return true;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.deQueue = function () {
    if (this.isEmpty()) {
        return false;
    }

    this.head = ++this.head % this.data.length;
    this.size--;

    return true;
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Front = function () {
    if (this.isEmpty()) {
        return -1;
    }

    return this.data[this.head];
};

/**
 * @return {number}
 */
MyCircularQueue.prototype.Rear = function () {
    if (this.isEmpty()) {
        return -1;
    }

    return this.data[this.tail];
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isEmpty = function () {
    return this.size == 0;
};

/**
 * @return {boolean}
 */
MyCircularQueue.prototype.isFull = function () {
    return this.size == this.data.length;
};

File: 2. Medium/622. Design Circular Queue/Solution.java
class MyCircularQueue {
    private int size = 0, head = 0, tail = -1;
    private final int[] arr;

    public MyCircularQueue(int k) {
        arr = new int[k];
    }

    public boolean enQueue(int value) {
        if (isFull())
            return false;

        tail = ++tail % arr.length;
        arr[tail] = value;
        size++;

        return true;
    }

    public boolean deQueue() {
        if (isEmpty())
            return false;

        head = ++head % arr.length;
        size--;

        return true;
    }

    public int Front() {
        return isEmpty() ? -1 : arr[head];
    }

    public int Rear() {
        return isEmpty() ? -1 : arr[tail];
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean isFull() {
        return size == arr.length;
    }
}